metomi/rose/loc_handlers/git.py

```diff
@@ -0,0 +1,178 @@
+
+    async def pull(self, loc, conf_tree):
+        """Get loc to its cache.
+
+        git sparse-checkout is not available below Git 2.25, and seems to omit
+        contents altogether if set to the root of the repo (as of 2.40.1).
+
+        Filtering requires uploadpack.allowFilter to be set true on
+        the remote repo or server.
+
+        """
+        if not loc.real_name:
+            self.parse(loc, conf_tree)
+        remote, path, ref = self._parse_name(loc)
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            git_dir_opt = f"--git-dir={tmpdirname}/.git"
+            await self.manager.popen.run_ok_async(
+                "git", git_dir_opt, "init"
+            )
+            if self.git_version >= (2, 25, 0) and path != "./":
+                # sparse-checkout available and suitable for this case.
+                await self.manager.popen.run_ok_async(
+                    "git", git_dir_opt, "sparse-checkout", "set", path,
+                    "--no-cone"
+                )
+                await self.manager.popen.run_ok_async(
+                    "git", git_dir_opt, "fetch", "--depth=1",
+                    "--filter=blob:none", remote, loc.key
+                )
+            else:
+                # Fallback.
+                await self.manager.popen.run_ok_async(
+                    "git", git_dir_opt, "fetch", "--depth=1", remote, loc.key
+                )
+
+            # Checkout to temporary location, then extract only 'path' later.
+            await self.manager.popen.run_ok_async(
+                "git", git_dir_opt, f"--work-tree={tmpdirname}", "checkout",
+                loc.key
+            )
+            name = tmpdirname + "/" + path
+
+            # Check that we have inferred the right type from the path name.
+            real_loc_type = (
+                loc.TYPE_TREE if os.path.isdir(name) else loc.TYPE_BLOB
+            )
+            if real_loc_type != loc.loc_type:
+                raise ValueError(
+                    f"Expected path '{path}' to be type '{loc.loc_type}', "
+                    + f"but it was '{real_loc_type}'. Check trailing slash."
+                )
+
+            # Extract only 'path' to cache.
+            dest = loc.cache
+            if loc.loc_type == "tree":
+                dest += "/"
+            cmd = self.manager.popen.get_cmd("rsync", name, dest)
+            await self.manager.popen.run_ok_async(*cmd)
+
+    def _parse_name(self, loc):
+        scheme, nonscheme = loc.name.split(":", 1)
+        return nonscheme.split(self.URI_SEPARATOR, maxsplit=3)
+
+    def _get_commithash(self, remote, ref):
+        """Get the commit hash given a branch, tag, or commit hash.
+
+        Short commit hashes will not resolve since there is no remote
+        rev-parse functionality.
+
+        """
+        ret_code, info, _ = self.manager.popen.run(
+            "git", "ls-remote", "--exit-code", remote, ref)
```

Comment:
According to the docs:

```
       --exit-code
           Exit with status "2" when no matching refs are found in the remote
           repository. Usually the command exits with status "0" to indicate it
           successfully talked with the remote repository, whether it found any
           matching refs.
```

Which suggests that it should be possible to separate "could not connect" from "no matching ref" by looking at the exit code?

Reply:
Looks like it works:

```console
$ git ls-remote --exit-code git@github.com:oliver-sanders/cylc-flow.git elephant; echo $?
2
$ git ls-remote --exit-code git@github.com:oliver-sanders/elephant.git master; echo $?
ERROR: Repository not found.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
128
```

Reply:
@benfitzpatrick - https://github.com/benfitzpatrick/rose/pull/18